File: RomanToInteger.cs
// Roman To Integer
// - Convert a Roman numeral string into an integer.
// - Valid symbols: I, V, X, L, C, D, M
// - Time: O(n), Space: O(1)

using System;
using System.Collections.Generic;

class Solution
{
    public int RomanToInt(string s)
    {
        var values = new Dictionary<char, int>()
        {
            {'I', 1},
            {'V', 5},
            {'X', 10},
            {'L', 50},
            {'C', 100},
            {'D', 500},
            {'M', 1000}
        };

        int total = 0;
        int prev = 0;

        for (int i = s.Length - 1; i >= 0; i--)
        {
            int curr = values[s[i]];
            if (curr < prev)
                total -= curr;
            else
                total += curr;
            prev = curr;
        }

        return total;
    }
}

File: RomanToInteger.py
# Roman To Integer
# - Convert a Roman numeral string into an integer.
# - Valid symbols: I, V, X, L, C, D, M
# - Constraints:
#   - 1 <= s.length <= 15
#   - s contains only valid Roman numerals.

class Solution:
    def romanToInt(self, s: str) -> int:
        values = {'I': 1, 'V': 5, 'X': 10, 'L': 50, 'C': 100, 'D': 500, 'M': 1000}
        total = 0
        prev = 0
        for char in reversed(s):
            curr = values[char]
            if curr < prev:
                total -= curr
            else:
                total += curr
            prev = curr
        return total
